scipy/stats/_distn_infrastructure.py

```diff
@@ -983,12 +983,19 @@ def rvs(self, *args, **kwds):
         if rndm is not None:
             self._random_state = random_state_saved
 
-        # Cast to int if discrete
+        # Cast to numpy.int64(vals is overflow) or int if discrete
         if discrete:
-            if size == ():
-                vals = int(vals)
+            if np.isscalar(vals):
```

Comment:
I don't think the scalar case makes sense here: `int` is arbitrary precision in python 3.

Reply:
not quite: it's used in `vals.astype(int)` below, which, IIUC, means "cast to a numpy integer of default width" (int32 on windows, int64 on linux).

That said, I restored the previous logic for the scalar / vector case. Possibly, the scalar case can be dropped. but I'm not sure and it's best done in a separate PR (which is easier to revert if needs be). The scary part is that `vals` comes from `vals = self._rvs(...)` above. Therefore, it's most likely a numpy array or scalar, i.e. it has the `astype` attribute. However, `_rvs` is for overriding by subclasses, meaning that in principle a subclass can return a python int, which does not have that attribute.

Reply:
I think scalar might be the wrong term, I meant the 0d array case covered by the original code.

The old code casts 0d arrays to `int` (`int(vals)` not `vals.astype(int)`) which has arbitrary precision in python 3 and so cannot overflow. Detecting large values and using `np.int64` instead doesn't make sense because it offers strictly _less_ precision.